spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java

```diff
@@ -172,25 +176,27 @@ public DefaultTaskExecutionService(LauncherRepository launcherRepository,
 	@Override
 	public long executeTask(String taskName, Map<String, String> taskDeploymentProperties, List<String> commandLineArgs,
 			String composedTaskRunnerName) {
-
+		// Get platform name and fallback to 'default'
 		String platformName = getPlatform(taskDeploymentProperties);
 
+		// Naive local state to prevent parallel launches to break things up
 		if(this.tasksBeingUpgraded.containsKey(taskName)) {
 			List<String> platforms = this.tasksBeingUpgraded.get(taskName);
 			if(platforms.contains(platformName)) {
-				throw new IllegalStateException(String.format("Unable to launch %s on platform %s because it is being upgraded", taskName, platformName));
+				throw new IllegalStateException(String.format(
+						"Unable to launch %s on platform %s because it is being upgraded", taskName, platformName));
 			}
 		}
 
-		// Remove since the key for task platform name will not pass validation for app, deployer, or scheduler prefix
+		// Remove since the key for task platform name will not pass validation for app,
+		// deployer, or scheduler prefix.
+		// Then validate
 		if (taskDeploymentProperties.containsKey(TASK_PLATFORM_NAME)) {
 			taskDeploymentProperties.remove(TASK_PLATFORM_NAME);
 		}
-
 		DeploymentPropertiesUtils.validateDeploymentProperties(taskDeploymentProperties);
 
-		TaskLauncher taskLauncher = findTaskLauncher(platformName);
-
+		// TODO: Same task name can only exist in one platform, um why?
```

Comment:
I think this is because our task execution model (and the corresponding batch model) is not aware of the platform.